File: mstar.py
import heapq
import itertools

from Mstar_cpp import python_phi

from graph import dijkstra_predecessor_and_distance


class Config:
    def __init__(self):
        self.cost = float('inf')
        self.heuristic = float('inf')
        self.collisions = set()
        self.back_set = []
        self.back_ptr = []

    def __str__(self):
        return str([self.cost, self.collisions, self.back_set, self.back_ptr])

    def __repr__(self):
        return str([self.cost, self.collisions, self.back_set, self.back_ptr])


class Mstar:
    def __init__(self, graph, v_I, v_W, v_F):
        self.n_agents = len(v_I)
        self.graph = graph
        self.v_I = v_I
        self.v_W = v_W
        for i in range(self.n_agents):
            start = v_I[i]
            end = v_F[i]
            if len(self.v_W[i]) > 0:
                self.v_W[i].sort(key=lambda x: euclidian_distance(start, x) / euclidian_distance(end, x))
        self.v_F = v_F
        self.configurations = {}
        self.policies = []
        self.distances = []
        self.targets = []
        for i in range(self.n_agents):
            policy_i = []
            distance_i = []
            targets_i = []
            for waypoint in v_W[i]:
                if waypoint in graph:
                    predecessor, distance = dijkstra_predecessor_and_distance(graph, waypoint)
                    policy_i.append(predecessor)
                    distance_i.append(distance)
                    targets_i.append(waypoint)
            predecessor, distance = dijkstra_predecessor_and_distance(graph, v_F[i])
            policy_i.append(predecessor)
            distance_i.append(distance)
            targets_i.append(v_F[i])
            self.policies.append(policy_i)
            self.distances.append(distance_i)
            self.targets.append(targets_i)
        self.open = []
        self.configurations[(v_I, (0,) * self.n_agents)] = Config()
        self.configurations[(v_I, (0,) * self.n_agents)].cost = 0
        heapq.heappush(self.open,
                       (self.heuristic_configuration(v_I, tuple([0] * self.n_agents)), (v_I, (0,) * self.n_agents)))

    def solve(self):
        configurations = self.configurations
        while len(self.open) > 0:
            v_k, target_indices = heapq.heappop(self.open)[1]
            v_k_config = configurations[(v_k, target_indices)]
            if v_k == self.v_F and all(target_indices[i] + 1 == len(self.targets[i]) for i in range(self.n_agents)):
                v_k_config.back_ptr.append(self.v_F)
                res = []
                for i in range(self.n_agents):
                    res.append([list(config[i]) for config in v_k_config.back_ptr])
                return res, v_k_config.cost
            if len(python_phi(v_k, [(-1, -1)])) == 0:
                for v_l in self.get_limited_neighbours(v_k, target_indices):
                    v_l_target_indices = list(target_indices)
                    for i in range(self.n_agents):
                        if v_l[i] == self.targets[i][v_l_target_indices[i]] and v_l[i] != self.v_F[i]:
                            v_l_target_indices[i] += 1
                    v_l_target_indices = tuple(v_l_target_indices)
                    v_l_collisions = python_phi(v_l, v_k)
                    if (v_l, v_l_target_indices) not in configurations:
                        configurations[(v_l, v_l_target_indices)] = Config()
                    v_l_config = configurations[(v_l, v_l_target_indices)]
                    v_l_config.collisions.update(v_l_collisions)
                    v_l_config.back_set.append((v_k, target_indices))
                    self.backprop(v_k, target_indices, v_l_config.collisions)

                    f = self.get_edge_weight(v_k, v_l, v_l_target_indices)
                    new_cost_v_l = v_k_config.cost + f
                    old_cost_v_l = v_l_config.cost
                    if len(v_l_collisions) == 0 and new_cost_v_l < old_cost_v_l:
                        v_l_config.cost = v_k_config.cost + f
                        v_l_config.back_ptr = v_k_config.back_ptr + [v_k]
                        heuristic = self.heuristic_configuration(v_l, v_l_target_indices)
                        heapq.heappush(self.open, (v_l_config.cost + heuristic, (v_l, v_l_target_indices)))
        return "No path exists, or I am an idiot"

    def get_limited_neighbours(self, v_k, target_indices):
        V_k = []
        options = []
        configurations = self.configurations
        for i in range(self.n_agents):
            vi_k = v_k[i]
            options_i = []
            if i in configurations[(v_k, target_indices)].collisions:
                # ADD all the neighbours
                options_i.append(vi_k)
                for nbr in self.graph[vi_k]:
                    options_i.append(nbr)
            else:
                target_index = target_indices[i]
                policy = self.policies[i][target_index]
                successors = policy[v_k[i]]
                if len(successors) == 0:
                    options_i.append(vi_k)
                else:
                    for successor in successors:
                        options_i.append(successor)
            options.append(options_i)
        if len(options) == 1:
            V_k.append((options[0][0],))
            return V_k
        for element in itertools.product(*options):
            V_k.append(element)
        return V_k

    def backprop(self, v_k, target_indices, C_l):
        v_k_config = self.configurations[(v_k, target_indices)]
        if not C_l.issubset(v_k_config.collisions):
            v_k_config.collisions.update(C_l)
            # Technically we should check whether its not already in open
            # But that takes too much time and it will settle it self
            # if not any(v_k in configuration for configuration in open):
            heuristic = self.heuristic_configuration(v_k, target_indices)
            heapq.heappush(self.open, (v_k_config.cost + heuristic, (v_k, target_indices)))
            for v_m, v_m_target_indices in v_k_config.back_set:
                self.backprop(v_m, v_m_target_indices, v_k_config.collisions)

    def get_edge_weight(self, v_k, v_l, target_indices):
        cost = 0
        for i in range(self.n_agents):
            if not (v_k[i] == v_l[i] == self.targets[i][-1] and target_indices[i] == len(self.targets[i]) - 1):
                cost += 1
        return cost

    def heuristic_configuration(self, v_k, target_indices):
        cost = 0
        for i in range(self.n_agents):
            target_index = target_indices[i]
            target = self.targets[i]
            cost += self.distances[i][target_index][v_k[i]]
            target_index += 1
            while target_index < len(target):
                cost += self.distances[i][target_index][target[target_index - 1]]
                target_index += 1
        return cost


def euclidian_distance(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

File: Cpp/Mstar/Mstar.cpp
#include <algorithm>
#include "Mstar.h"
#include "inc/Graph.h"
#include "inc/Utils.h"

Mstar::Mstar(std::vector<std::vector<int>>& grid, std::vector<std::pair<int, int>>& v_I_a,
             std::vector<std::vector<std::pair<int, int>>>& v_W_a,
             std::vector<std::pair<int, int>>& v_F_a) {
    n_agents = v_I_a.size();
    graph = create_graph(grid);
    for (std::pair<int, int> &start : v_I_a) {
        v_I.emplace_back(start.first, start.second);
    }
    for (std::pair<int, int> &goal : v_F_a) {
        v_F.emplace_back(goal.first, goal.second);
    }
    for (const std::vector<std::pair<int, int>> &waypoints : v_W_a) {
        std::vector<Coordinate> waypoints_i;
        for (std::pair<int, int> waypoint : waypoints) {
            if (waypoint.first != -1 && waypoint.second != -1) {
                waypoints_i.emplace_back(waypoint.first, waypoint.second);
            }
        }
        v_W.push_back(waypoints_i);
    }
    update_policies_distances_targets();
    bool optimal = true;
    for (int i = 0; i != n_agents; i++) {
        if (optimal) {
            v_W[i] = tsp_dynamic(v_I[i], v_F[i], v_W[i], distances[i]);
        } else {
            v_W[i] = tsp_greedy(v_I[i], v_F[i], v_W[i], distances[i]);
        }
    }
    update_policies_distances_targets();
    Config_key v_I_key = Config_key(v_I, std::vector<int>(n_agents, 0));
    configurations[v_I_key] = Config_value();
    configurations[v_I_key].cost = 0;
    open.push(Queue_entry(heuristic_configuration(v_I_key), v_I_key));
}

void Mstar::update_policies_distances_targets() {
    policies = {};
    distances = {};
    targets = {};
    for (int i = 0; i != n_agents; i++) {
        std::vector<std::unordered_map<Coordinate, std::vector<Coordinate>, coordinate_hash>> policy_i;
        std::vector<std::unordered_map<Coordinate, int, coordinate_hash>> distance_i;
        std::vector<Coordinate> targets_i;
        for (Coordinate const &waypoint : v_W[i]) {
            if (graph.count(waypoint)) {
                std::pair<std::unordered_map<Coordinate, std::vector<Coordinate>, coordinate_hash>, std::unordered_map<Coordinate, int, coordinate_hash>> dijkstra = dijkstra_predecessor_and_distance(
                        graph, waypoint);
                policy_i.push_back(dijkstra.first);
                distance_i.push_back(dijkstra.second);
                targets_i.push_back(waypoint);
            }
        }
        std::pair<std::unordered_map<Coordinate, std::vector<Coordinate>, coordinate_hash>, std::unordered_map<Coordinate, int, coordinate_hash>> dijkstra = dijkstra_predecessor_and_distance(
                graph, v_F[i]);
        policy_i.push_back(dijkstra.first);
        distance_i.push_back(dijkstra.second);
        targets_i.push_back(v_F[i]);
        policies.push_back(policy_i);
        distances.push_back(distance_i);
        targets.push_back(targets_i);
    }
}

std::pair<std::vector<std::vector<std::pair<int, int>>>, int> Mstar::solve() {
    while (!open.empty()) {
        Config_key v_k = open.top().config_key;
        std::cout << v_k << std::endl;
        Config_value v_k_config = configurations[v_k];
        open.pop();
        if (v_k.coordinates == v_F) {
            bool visited_waypoints = true;
            for (int i = 0; i != n_agents; i++) {
                if (v_k.targets[i] != targets[i].size() - 1) {
                    visited_waypoints = false;
                    break;
                }
            }
            if (visited_waypoints) {
                v_k_config.back_ptr.push_back(v_F);
                std::vector<std::vector<std::pair<int, int>>> res;
                for (int i = 0; i != n_agents; i++) {
                    std::vector<std::pair<int, int>> res_i;
                    for (auto &config : v_k_config.back_ptr) {
                        res_i.emplace_back(config[i].a, config[i].b);
                    }
                    res.push_back(res_i);
                }
                std::cout << "Get edge weight took " << get_edge_weight_time << std::endl;
                std::cout << "Backrpop took " << backprop_time << std::endl;
                std::cout << "Get limited neighbours took " << get_limited_neighbours_time << std::endl;
                std::cout << "Heuristic configuration took " << heuristic_configuration_time << std::endl;
                std::cout << "Phi took " << phi_time << std::endl;
                return make_pair(res, v_k_config.cost);
            }
        }
        for (std::vector<Coordinate> &v_l : get_limited_neighbours(v_k)) {
            std::vector<int> v_l_target_indices = v_k.targets;
            for (int i = 0; i != n_agents; i++) {
                if (v_l[i] == targets[i][v_l_target_indices[i]] && v_l[i] != v_F[i]) {
                    v_l_target_indices[i]++;
                }
            }
            time.start();
            std::set<int> v_l_collisions = phi(v_l, v_k.coordinates);
            time.stop();
            phi_time += time.elapsed();
            Config_key v_l_key = Config_key(v_l, v_l_target_indices);
            Config_value v_l_config;
            if (!configurations.count(v_l_key)) {
                v_l_config = Config_value();
            } else {
                v_l_config = configurations[v_l_key];
            }
            v_l_config.collisions.insert(v_l_collisions.begin(), v_l_collisions.end());
            v_l_config.back_set.insert(v_k);
            time.start();
            backprop(v_k, v_l_config.collisions);
            time.stop();
            backprop_time += float(time.elapsed());
            time.start();
            int f = get_edge_weight(v_k, v_l_key);
            time.stop();
            get_edge_weight_time += float(time.elapsed());
            int new_cost_v_l = v_k_config.cost + f;
            int old_cost_v_l = v_l_config.cost;
            if (v_l_collisions.empty() && new_cost_v_l < old_cost_v_l) {
                v_l_config.cost = new_cost_v_l;
                v_l_config.back_ptr = v_k_config.back_ptr;
                v_l_config.back_ptr.push_back(v_k.coordinates);
                configurations[v_l_key] = v_l_config;
                int heuristic = heuristic_configuration(v_l_key);
                open.push(Queue_entry(v_l_config.cost + heuristic, v_l_key));
            }
        }
    }
    std::cout << "No path exists, or I am an idiot";
}

int Mstar::get_edge_weight(Config_key &v_k, Config_key &v_l) {
    int cost = 0;
    for (int i = 0; i != n_agents; i++) {
        if (!(v_k.coordinates[i] == v_l.coordinates[i] && v_k.coordinates[i] == v_F[i] &&
              v_l.targets[i] == targets[i].size() - 1)) {
            cost++;
        }
    }
    return cost;
}

void Mstar::backprop(Config_key &v_k, std::unordered_set<int> &C_l) {
    Config_value &current_config = configurations[v_k];
    bool subset = isSubset(C_l, current_config.collisions);
    if (!subset) {
        current_config.collisions.insert(C_l.begin(), C_l.end());
        int heuristic = heuristic_configuration(v_k);

        open.push(Queue_entry(current_config.cost + heuristic, v_k));
        for (Config_key v_m : current_config.back_set) {
            backprop(v_m, current_config.collisions);
        }
    }
}

std::vector<std::vector<Coordinate>> Mstar::get_limited_neighbours(Config_key &v_k) {
    time.start();
    std::vector<std::vector<Coordinate>> options;
    for (int i = 0; i != n_agents; i++) {
        Coordinate source = v_k.coordinates[i];
        std::vector<Coordinate> options_i;
        if (configurations[v_k].collisions.count(i)) {
            // Add all the neighbours
            options_i.push_back(source);
            int target_index = v_k.targets[i];
            std::vector<Coordinate> successors = policies[i][target_index][source];
            if (successors.size() > 1) {
                for (auto &successor : successors) {
                    options_i.push_back(successor);
                }
            } else {
                for (auto &nbr : graph[source]) {
                    options_i.push_back(nbr);
                }
            }
        } else {
            int target_index = v_k.targets[i];
            std::vector<Coordinate> successors = policies[i][target_index][source];
            if (successors.empty()) {
                options_i.push_back(source);
            } else {
                options_i.push_back(successors[0]);
            }
        }
        options.push_back(options_i);
    }
    std::vector<std::vector<Coordinate>> res;
    if (options.size() == 1) {
        res = options;
    } else {
        res = cart_product(options);
    }
    time.stop();
    get_limited_neighbours_time += time.elapsed();
    return res;
}

int Mstar::heuristic_configuration(Config_key &v_k) {
    time.start();
    int cost = 0;
    for (int i = 0; i != n_agents; i++) {
        int target_index = v_k.targets[i];
        std::vector<Coordinate> targets_i = targets[i];
        cost += distances[i][target_index][v_k.coordinates[i]];
        target_index++;
        while (target_index < targets_i.size()) {
            cost += distances[i][target_index][targets_i[target_index - 1]];
            target_index++;
        }
    }
    time.stop();
    heuristic_configuration_time += time.elapsed();
    return cost;
}